feat!: datetimepicker validation improvements

ugur-vaadin opened the pull request:
## Description

- Adds `unparsable-change` event for the web component.
- Handles the case where one of the fields is still empty. Validates if
  - the new value in the edited field is unparsable
  - the edited field is date picker and the date is out of range set using `min` and/or `max`
- improves validation consistency. Except for cases where the the edited field is valid after the edit and the `DateTimePicker` is still being edited, the validation behaves the same for Enter, outside click, and blur.
- Adds tests that verify the validation logic and change events using combinations of initial states and field updates.

Based on the [prototype implementation](https://github.com/vaadin/web-components/tree/proto/date-time-picker/unparsable-incomplete-change).

Part of [#6697](https://github.com/vaadin/flow-components/issues/6697).

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/pr
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.
- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

vursen commented on packages/date-time-picker/test/individual-picker-update-helpers.js:
```diff
@@ -0,0 +1,94 @@
+};
+
+export const individualPickerUpdateTestSetup = (individualPickerTest) => {
+  ['date-picker', 'time-picker'].forEach((pickerType) => {
```
Would it be possible to simplify or restructure these tests a bit to improve readability? The deep nesting is making it tricky to understand the test cases. Maybe it would be easier to follow if the tests were more explicit, even with some duplication.

ugur-vaadin replied:
I started with that approach, but it seemed to bloat the tests by quite a lot. But I don't have a strong opinion, and if you think it is hard to read I can change it. Maybe we can keep `['date-picker', 'time-picker'].forEach(...)` and explicitly add tests for different states&updates. What do you think?

vursen replied:
> Maybe we can keep ['date-picker', 'time-picker'].forEach(...) and explicitly add tests for different states&updates

Sounds good! Let's try this.

ugur-vaadin replied:
Separated the tests.

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker.d.ts:
```diff
@@ -93,7 +93,24 @@ export interface DateTimePickerEventMap extends DateTimePickerCustomEventMap, HT
+ *
+ * Value change             | Event
+ * :------------------------|:------------------
+ * empty => parsable        | change
+ * empty => unparsable      | unparsable-change
+ * parsable => empty        | change
+ * parsable => parsable     | change
+ * parsable => unparsable   | change
+ * unparsable => empty      | unparsable-change
+ * unparsable => parsable   | change
+ * unparsable => unparsable | unparsable-change
```
I'd suggest including the `incomplete` scenario in this list separately.

ugur-vaadin replied:
Done.

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker.d.ts:
```diff
@@ -93,7 +93,24 @@ export interface DateTimePickerEventMap extends DateTimePickerCustomEventMap, HT
+ * unparsable => unparsable | unparsable-change
+ *
  * @fires {Event} change - Fired when the user commits a value change.
+ * @fires {Event} unparsable-change - Fired when the user commits an unparsable value change and there is no change event.
```
```suggestion
 * @fires {Event} unparsable-change - Fired when the user commits an unparsable or incomplete value change and there is no change event.
```

vursen commented on packages/date-time-picker/test/validation.test.js:
```diff
@@ -60,16 +60,395 @@ const fixtures = {
     });
 
-    it('should validate on date-picker blur', () => {
+    ['date-picker', 'time-picker'].forEach((pickerType) => {
```
Let's create a `value-commit.test.js` file and move all new tests there. Validation and change events are both part of the value commit flow, so testing them together in one place would make sense. Example:

https://github.com/vaadin/web-components/blob/7c2b6e54be7471f6e34b6c7495f3dfd111fe0d3e/packages/date-picker/test/value-commit.test.js#L12-L43

ugur-vaadin replied:
Done.

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker-mixin.js:
```diff
@@ -724,8 +804,25 @@ export const DateTimePickerMixin = (superClass) =>
+      this.__committedUnparsableValue = this.__unparsableValue;
+    }
+
+    get __hasPendingValueChange() {
```
```suggestion
    /** @private */
    get __hasPendingValueChange() {
```

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker-mixin.js:
```diff
@@ -315,12 +316,37 @@ export const DateTimePickerMixin = (superClass) =>
       return [this.__datePicker, this.__timePicker];
     }
 
+    get __filledPickers() {
```
```suggestion
    /** @private */
    get __filledPickers() {
```

vursen commented:
Looks like this [change](https://github.com/vaadin/web-components/pull/8986/files#diff-030d516be87abc974f9dfb16537b8b08f8ca8bdd1e84a655173dac172d6cbf98R430) and this [change](https://github.com/vaadin/web-components/pull/8986/files#diff-030d516be87abc974f9dfb16537b8b08f8ca8bdd1e84a655173dac172d6cbf98R496) together break required validation on blur, which works in other components:

https://github.com/user-attachments/assets/2bd0cb85-24f6-45e8-ba7e-fdcd9dbc3349

ugur-vaadin commented:
> Looks like this [change](https://github.com/vaadin/web-components/pull/8986/files#diff-030d516be87abc974f9dfb16537b8b08f8ca8bdd1e84a655173dac172d6cbf98R430) and this [change](https://github.com/vaadin/web-components/pull/8986/files#diff-030d516be87abc974f9dfb16537b8b08f8ca8bdd1e84a655173dac172d6cbf98R496) together break required validation on blur, which works in other components:
> 
>  Screen.Recording.2025-04-29.at.14.36.43.mov

I think that is my misunderstanding rather than an implementation error. My understanding was that the validation should not run when the component is blurred with no change to the input even if it is required.

vursen commented:
> My understanding was that the validation should not run when the component is blurred with no change to the input even if it is required.

Yeah, there is a difference in component behavior: web components validate on blur, whereas flow components do not.

vursen commented:
> This was fixed and the tests for validating on blur were reverted.

This [place](https://github.com/vaadin/web-components/pull/8986/files#diff-030d516be87abc974f9dfb16537b8b08f8ca8bdd1e84a655173dac172d6cbf98R496) needs to be fixed too, as otherwise required validation doesn't occur when the overlay is closed by clicking outside – a case that is also treated as blur in other components such as date-picker.

ugur-vaadin commented:
> > This was fixed and the tests for validating on blur were reverted.
> 
> I guess this [place](https://github.com/vaadin/web-components/pull/8986/files#diff-030d516be87abc974f9dfb16537b8b08f8ca8bdd1e84a655173dac172d6cbf98R496) needs to be fixed too, as otherwise required validation doesn't occur when the overlay is closed by clicking outside – a case that is also treated as blur in other components such as date-picker, even though focus is restored afterwards.

Fixed and added tests.

ugur-vaadin commented on packages/date-time-picker/test/value-commit.test.js:
```diff
@@ -0,0 +1,567 @@
+    it('should validate on time-picker blur', () => {
+      timePicker.focus();
+      timePicker.blur();
+      expect(validateSpy.calledOnce).to.be.true;
```
Refactored the tests.

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker-mixin.js:
```diff
@@ -347,12 +376,49 @@ export const DateTimePickerMixin = (superClass) =>
+      if (isOutsideClick) {
+        this.__outsideClickInProgress = true;
+
+        setTimeout(() => {
```
Could we use `queueMicrotask` here to limit more preciesely how long this flag stays true?

ugur-vaadin replied:
It seems that `queueMicrotask` does not provide enough of a duration for the flag to stay true.

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker-mixin.js:
```diff
@@ -347,12 +376,49 @@ export const DateTimePickerMixin = (superClass) =>
+    /** @protected */
+    connectedCallback() {
+      super.connectedCallback();
+      document.documentElement.addEventListener('click', this.__onGlobalClick, true);
```
```suggestion
      document.addEventListener('click', this.__onGlobalClick, true);
```

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker-mixin.js:
```diff
@@ -347,12 +376,49 @@ export const DateTimePickerMixin = (superClass) =>
+    /** @protected */
+    disconnectedCallback() {
+      super.disconnectedCallback();
+      document.documentElement.removeEventListener('click', this.__onGlobalClick, true);
```
```suggestion
      document.removeEventListener('click', this.__onGlobalClick, true);
```

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker-mixin.js:
```diff
@@ -408,33 +474,55 @@ export const DateTimePickerMixin = (superClass) =>
 
+    /** @private */
+    __requestValueCommit() {
+      this.__valueCommitDebouncer = Debouncer.debounce(this.__valueCommitDebouncer, microTask, () => {
```
Ideally, we should avoid changing the timing of validation to ensure that the updated invalid state is already available in user-defined blur listeners. This change seems reasonable, but I'm wondering if it should instead be made synchronous. I think we could just let validation to run multiple times for now and fix it later, as the same issue exists in other components as well.

ugur-vaadin replied:
Reverted the change.

ugur-vaadin commented on packages/date-time-picker/test/value-commit.test.js:
```diff
@@ -0,0 +1,373 @@
+
+  async function expectNoValueCommit() {
+    await nextRender();
+    expect(validateSpy.called).to.be.false;
```
Done.

vursen commented on packages/date-time-picker/src/vaadin-date-time-picker-mixin.js:
```diff
@@ -408,33 +472,53 @@ export const DateTimePickerMixin = (superClass) =>
 
+    /** @private */
+    __requestValueCommit() {
+      this._requestValidation();
```
It seems that `this._requestValidation()` could be now called from `this.__commitPendingValueChange()`, removing the need for this extra method.

ugur-vaadin replied:
Done.

vursen commented on packages/date-time-picker/test/value-commit.test.js:
```diff
@@ -0,0 +1,376 @@
+    it('should commit on date-picker outside click when value is changed to a date inside the set range while time-picker is filled', async () => {
+      await enterParsableInput(timePicker);
+      await enterInput(datePicker, '2/2/1985');
+      resetSpyHistories();
```
Just entering input shouldn't result in a commit. The same applies to other tests in this suite.

```suggestion
```

ugur-vaadin replied:
Removed the reset from the tests.

vursen commented on packages/date-time-picker/test/value-commit.test.js:
```diff
@@ -0,0 +1,376 @@
+      });
+    });
+  });
+  describe('initially invalid value', () => {
```
Let's also add tests for an initially invalid empty field, which can take place when a required datetimepicker is blurred without any input. This is the opposite scenario to `initially invalid value`.

```js
describe('initially invalid empty field', () => {
  beforeEach(() => {
    dateTimePicker.required = true;
    dateTimePicker.validate();
  });

  describeForEachPicker('incomplete input entered', () => {
     // ...
  });
})
```

vursen commented on packages/date-time-picker/test/value-commit.test.js:
```diff
@@ -0,0 +1,430 @@
+      });
+
+      it('should commit as unparsable on Enter', async () => {
+        await enterParsableInput(picker);
```
Let's move this to `beforeEach`

```suggestion
        await enterParsableInput(picker);
```

ugur-vaadin replied:
Moved.

vursen commented on packages/date-time-picker/test/value-commit.test.js:
```diff
@@ -0,0 +1,430 @@
+      });
+
+      it('should commit as unparsable on Enter', async () => {
+        await enterUnparsableInput(picker);
```
Let's move this to `beforeEach`

ugur-vaadin replied:
Moved.
